### PR TITLE
L-BFGS solver based on scipy.optimize with L2 regularization

### PR DESCRIPTION
**Pull request by MLnick**
Refer to #30 for the discussion and prototype from which this was developed.

Still a little bit WIP, though it supports `L2` regularizer and:
* is equivalent to sklearn's `LogisticRegression` with `solver=lbfgs` and `fit_intercept=False`
* doesn't give precisely the same solution as sklearn's `Ridge` for squared loss (as the solvers are different), but achieves essentially the same RMSE metric on some simple test data (`fit_intercept=False`)

Todos:
- [ ] test cases
- [ ] maybe some performance testing (refer e.g. https://gist.github.com/mrocklin/4e486064882cce630ffb4ee4e39bc333)
- [ ] support `L1` regularizer (via adapting https://gist.github.com/vene/fab06038a00309569cdd). Could this potentially also support `ElasticNet` regularizer?

**Note** not sure if the estimator versions will give the same results as sklearn because intercept handling may be different. Still need to investigate.

### CODE REVIEW

**Comment by MLnick** [comment]
cc @mrocklin @moody-marlin @TomAugspurger

**Comment by MLnick** [comment]
Script to produce the sklearn comparisons:

```python
import dask
import dask.array as da
from dask_glm.algorithms import lbfgs
from dask_glm.families import Logistic, Normal
import numpy as np
from sklearn import datasets
from sklearn.linear_model import LogisticRegression, Ridge
from sklearn.metrics import roc_auc_score, mean_squared_error

# compare with sklearn's LogisticRegression using L-BFGS solver
# make dataset
X, y = datasets.make_classification(n_classes=2, n_samples=10000, n_features=100)
# dask versions
Xda = da.from_array(X, 1000)
yda = da.from_array(y, 1000)

lr = LogisticRegression(fit_intercept=False, solver="lbfgs", max_iter=100)

betas_sklearn = []
betas_dask = []
for c in [1e5, 10.0, 1.0, 0.01]:
	lr.C = c
	lr.fit(X, y)
	betas_sklearn.append(lr.coef_)
	alpha = 1/c
	dask_beta = lbfgs(Xda, yda, alpha=alpha, max_iter=100, family=Logistic)
	betas_dask.append(dask_beta)

assert np.alltrue([np.allclose(a, b) for a, b in zip(betas_sklearn, betas_dask)]), "Coefficients for Logistic differ between sklearn and dask-glm"

# compare with sklearn's Ridge regressor
# make dataset
Xr, yr = datasets.make_regression(n_samples=10000, n_features=100)
# dask versions
Xrda = da.from_array(Xr, 1000)
yrda = da.from_array(yr, 1000)
ridge = Ridge(fit_intercept=False)

betas_sklearn_ridge = []
betas_dask_ridge= []
for l in [1e-10, 1e-3, 1e-1, 1.0, 10]:
	ridge.alpha = l
	ridge.fit(Xr, yr)
	betas_sklearn_ridge.append(ridge.coef_)
	dask_beta = lbfgs(Xrda, yrda, alpha=l, max_iter=100, family=Normal)
	betas_dask_ridge.append(dask_beta)

	ridge2 = Ridge(fit_intercept=False)
	ridge2.coef_ = dask_beta
	ridge2.intercept_ = 0.0
	print("alpha=%2.10f, RMSE Ridge=%2.6f, RMSE dask-glm=%2.6f" % (l, np.sqrt(mean_squared_error(y, ridge.predict(Xr))), np.sqrt(mean_squared_error(y, ridge2.predict(Xr)))))

```

File: dask_glm/algorithms.py [hunk 1]
```diff
@@ -243,78 +243,31 @@ def shrinkage(x, kappa):
...
-
-        beta = beta - stepSize * step  # tiny bit of repeat work here to avoid communication
-        Xbeta = Xbeta - stepSize * Xstep
-
-        if stepSize == 0:
-            print('No more progress')
-            break
+    pointwise_loss = family.pointwise_loss
+    pointwise_gradient = family.pointwise_gradient
+    regularizer = Regularizer.get('l2')
```
Comment (moody-marlin) [inline dask_glm/algorithms.py]: Could you elevate `regularizer` to a keyword?  On first glance it seems like this would work perfectly fine with `regularizer=None` or any other regularizer that is differentiable.  If the regularizer classes implement their methods appropriately, then if an algorithm runs into a point of non-differentiability, an error is thrown.

**Comment by mrocklin** [comment]
I've taken the liberty of adding `lbfgs` to the testing suite in https://github.com/dask/dask-glm/pull/50/commits/8c1345b76af1d203d30aab33ee80365b732af5be .  Hope you don't mind me pushing to your fork @MLnick .  

It looks like most things work well, a couple things are failing tolerance checks.

**Comment by mrocklin** [comment]
I ran this on a slightly larger problem with the distributed scheduler (just on my local laptop) and got [this profile](https://rawgit.com/mrocklin/d27442d02c0edfa5e1a049beada00a4c/raw/b063fdda6d3029fef4abbc064ba610556eef041c/task-stream-lbfgs.html).  I recommend clicking the wheel zoom tool and then zooming into an arbitrary part of the computation.

We're still mostly bound by whitespace / dead time.  We can resolve a lot of this just by using larger chunks, but this does motivate further work in optimizations.  I think that we could dramatically reduce overhead here if we did two things:

1.  Fused with a wider width (something like 10)
2.  Produced the graph only once, rather than at every iteration

You've correctly turned off fusion for the computation and I have no doubt that this will speed things up, especially on small chunks.  We should make it easier to produce the graph once though.  I think that this will allow us to be much more aggressive about optimization in the future.  This is something that I probably need to do within dask proper.

After this it looks like we're just bound by tensordot and exp, which is pretty typical.

Anyway, nothing for you to do here, I just thought you might enjoy seeing the profile plot and hearing about possibilities to make this faster in the future.

**Comment by moody-marlin** [comment]
@mrocklin it looks like `lbfgs` is currently failing the tolerance checks just because `l2`-regularization is hard-coded into it and the tests are testing for unregularized problems; should be a quick fix.

**Comment by MLnick** [comment]
@mrocklin thanks for pushing the test integration. For the fusion I just followed your example notebook referenced in #30 where you mentioned it would be better for performance.

@moody-marlin I can make `regularizer` a keyword arg, sure.

**Comment by MLnick** [comment]
I still see some tolerance-related test failures locally in `dask_glm/tests/test_algos_families.py`/`test_methods`

**Comment by moody-marlin** [comment]
@MLnick it looks like `test_methods` is failing -- this test is testing for accuracy on algorithms which it assumes are defaulting to solving unregularized problems.  If you change the default of `lbfgs` to `regularizer=None` (and update [these lines](https://github.com/MLnick/dask-glm/blob/dfdefdd1f6644077c6dc793b6f3c64900b90eeb1/dask_glm/algorithms.py#L258-L259) to not do anything) I expect the test to pass.

This raises a good issue though - we might need to update the tests to inspect for a `regularizer` kwarg and adapt accordingly.

**Comment by mrocklin** [comment]
> This raises a good issue though - we might need to update the tests to inspect for a regularizer kwarg and adapt accordingly.

We could add in kwargs to the pytest parameters along with the algorithm like we do in some other tests.

**Comment by moody-marlin** [comment]
Yea that could work well; the only added complexity is that there are some algorithm / regularizer pairs that won't work together (example: `newton` and `l1`) so we would need to handle that appropriately.

**Comment by MLnick** [comment]
@moody-marlin ah right. I agree it makes most sense to set the regularizer
default to None and adapt the lbfgs code accordingly.

**Comment by MLnick** [comment]
@mrocklin by the way, I keep getting the following warning for `lbfgs`: 

```
.../dask/array/core.py:476: RuntimeWarning: overflow encountered in true_divide
```

**Comment by mrocklin** [comment]
Looks like tests are passing :)

**Comment by MLnick** [comment]
Do we need any further tests?

**Comment by moody-marlin** [comment]
I think we're good on tests for now but we should continue to think about how we can make them more comprehensive and leave that for a future PR.

LGTM!

**Comment by mrocklin** [comment]
Merged.  Thanks @MLnick !  I'm looking forward to playing with this.

**Comment by MLnick** [comment]
Awesome, thanks!

By the way any thoughts on changing "lamduh" -> "alpha" for consistency
with e.g. sklearn? :)

**Comment by MLnick** [comment]
I will also try to do some larger-scale testing on my cluster :)
